include/nlohmann/json.hpp

```diff
@@ -3486,31 +3489,32 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
     access by reference
     @sa see @ref value() for access by value with a default value
 
-    @since version 1.0.0
+    @since version 1.0.0; template added in version 3.10.0
 
     @liveexample{The example below shows how object elements can be read and
     written using `at()`. It also demonstrates the different exceptions that
     can be thrown.,at__object_t_key_type}
     */
-    reference at(const typename object_t::key_type& key)
+    template < class KeyT, typename std::enable_if <
+                   !std::is_same<typename std::decay<KeyT>::type, json_pointer>::value&& (
+#if defined(JSON_HAS_CPP_17)
+                       std::is_same<typename std::decay<KeyT>::type, std::string_view>::value ||
+#endif
+                       std::is_convertible<KeyT, typename object_t::key_type>::value), int >::type    = 0 >
+    reference at(KeyT && key)
     {
         // at only works for objects
-        if (JSON_HEDLEY_LIKELY(is_object()))
+        if (JSON_HEDLEY_UNLIKELY(!is_object()))
         {
-            JSON_TRY
-            {
-                return set_parent(m_value.object->at(key));
-            }
-            JSON_CATCH (std::out_of_range&)
-            {
-                // create better exception explanation
-                JSON_THROW(out_of_range::create(403, "key '" + key + "' not found", *this));
-            }
+            JSON_THROW(type_error::create(304, "cannot use at() with " + std::string(type_name()), *this));
```

Comment:
Probably you want

static constexpr auto msgPrefix = "cannot use at() with ";
auto const& typeName = type_name();
std::string exceptionMsg;
exceptionMsg.reserve(std::size(msgPrefix) + typeName.size());
exceptionMsg.append(msgPRefix);
exceptionMsg.append(typeName);
JSON_THROW(type_error::create(304, std::move(exceptionMsg), *this));

This saves you at least 2 std::string constructions.

Reply:
Or if you want something a bit less yucky to type out, then something like this:

https://github.com/splinter-build/splinter/blob/15-default-values/src/string_concat.h

Reply:
I'd rather not do this in the scope of this PR.

Reply:
Understood.